// @socket.io/redis-streams-adapter/dist/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PUBSUB = exports.SSUBSCRIBE = exports.SUBSCRIBE = exports.SPUBLISH = exports.PUBLISH = exports.hashCode = exports.GETDEL = exports.SET = exports.XRANGE = exports.XADD = exports.XREAD = exports.duplicateClient = exports.hasBinary = void 0;
function hasBinary(obj, toJSON) {
    if (!obj || typeof obj !== "object") {
        return false;
    }
    if (obj instanceof ArrayBuffer || ArrayBuffer.isView(obj)) {
        return true;
    }
    if (Array.isArray(obj)) {
        for (let i = 0, l = obj.length; i < l; i++) {
            if (hasBinary(obj[i])) {
                return true;
            }
        }
        return false;
    }
    for (const key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key) && hasBinary(obj[key])) {
            return true;
        }
    }
    if (obj.toJSON && typeof obj.toJSON === "function" && !toJSON) {
        return hasBinary(obj.toJSON(), true);
    }
    return false;
}
exports.hasBinary = hasBinary;
/**
 * Whether the client comes from the `redis` package
 *
 * @param redisClient
 *
 * @see https://github.com/redis/node-redis
 */
function isRedisV4Client(redisClient) {
    return typeof redisClient.sSubscribe === "function";
}
async function duplicateClient(redisClient) {
    const newClient = redisClient.duplicate();
    newClient.on("error", (err) => {
        // ignore errors
    });
    if (isRedisV4Client(redisClient) || redisClient.options.lazyConnect) {
        await newClient.connect();
    }
    return newClient;
}
exports.duplicateClient = duplicateClient;
/**
 * Map the output of the XREAD/XRANGE command with the ioredis package to the format of the redis package
 * @param result
 */
function mapResult(result) {
    const id = result[0];
    const inlineValues = result[1];
    const message = {};
    for (let i = 0; i < inlineValues.length; i += 2) {
        message[inlineValues[i]] = inlineValues[i + 1];
    }
    return {
        id,
        message,
    };
}
/**
 * @see https://redis.io/commands/xread/
 */
function XREAD(redisClient, streamName, offset, readCount, blockTimeInMs) {
    if (isRedisV4Client(redisClient)) {
        return redisClient.xRead([
            {
                key: streamName,
                id: offset,
            },
        ], {
            COUNT: readCount,
            BLOCK: blockTimeInMs,
        });
    }
    else {
        return redisClient
            .xread("BLOCK", blockTimeInMs, "COUNT", readCount, "STREAMS", streamName, offset)
            .then((results) => {
            if (results === null) {
                return null;
            }
            return [
                {
                    messages: results[0][1].map(mapResult),
                },
            ];
        });
    }
}
exports.XREAD = XREAD;
/**
 * @see https://redis.io/commands/xadd/
 */
function XADD(redisClient, streamName, payload, maxLenThreshold) {
    if (isRedisV4Client(redisClient)) {
        return redisClient.xAdd(streamName, "*", payload, {
            TRIM: {
                strategy: "MAXLEN",
                strategyModifier: "~",
                threshold: maxLenThreshold,
            },
        });
    }
    else {
        const args = [streamName, "MAXLEN", "~", maxLenThreshold, "*"];
        Object.keys(payload).forEach((k) => {
            args.push(k, payload[k]);
        });
        return redisClient.xadd.call(redisClient, args);
    }
}
exports.XADD = XADD;
/**
 * @see https://redis.io/commands/xrange/
 */
function XRANGE(redisClient, streamName, start, end) {
    if (isRedisV4Client(redisClient)) {
        return redisClient.xRange(streamName, start, end);
    }
    else {
        return redisClient.xrange(streamName, start, end).then((res) => {
            return res.map(mapResult);
        });
    }
}
exports.XRANGE = XRANGE;
/**
 * @see https://redis.io/commands/set/
 */
function SET(redisClient, key, value, expiryInSeconds) {
    if (isRedisV4Client(redisClient)) {
        return redisClient.set(key, value, {
            PX: expiryInSeconds,
        });
    }
    else {
        return redisClient.set(key, value, "PX", expiryInSeconds);
    }
}
exports.SET = SET;
/**
 * @see https://redis.io/commands/getdel/
 */
function GETDEL(redisClient, key) {
    if (isRedisV4Client(redisClient)) {
        // note: GETDEL was added in Redis version 6.2
        return redisClient.multi().get(key).del(key).exec();
    }
    else {
        return redisClient
            .multi()
            .get(key)
            .del(key)
            .exec()
            .then((res) => {
            return [res[0][1]];
        });
    }
}
exports.GETDEL = GETDEL;
function hashCode(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        let chr = str.charCodeAt(i);
        hash = hash * 31 + chr;
        hash |= 0;
    }
    return hash;
}
exports.hashCode = hashCode;
function PUBLISH(redisClient, channel, payload) {
    return redisClient.publish(channel, payload);
}
exports.PUBLISH = PUBLISH;
function SPUBLISH(redisClient, channel, payload) {
    if (isRedisV4Client(redisClient)) {
        return redisClient.sPublish(channel, payload);
    }
    else {
        return redisClient.spublish(channel, payload);
    }
}
exports.SPUBLISH = SPUBLISH;
const RETURN_BUFFERS = true;
function SUBSCRIBE(subClient, channels, listener) {
    if (isRedisV4Client(subClient)) {
        subClient.subscribe(channels, listener, RETURN_BUFFERS);
    }
    else {
        subClient.subscribe(channels);
        subClient.on("messageBuffer", (channel, payload) => {
            if (channels.includes(channel.toString())) {
                listener(payload);
            }
        });
    }
}
exports.SUBSCRIBE = SUBSCRIBE;
function SSUBSCRIBE(subClient, channels, listener) {
    if (isRedisV4Client(subClient)) {
        // note: we could also have used a hash tag ({...}) to ensure the channels are mapped to the same slot
        for (const channel of channels) {
            subClient.sSubscribe(channel, listener, RETURN_BUFFERS);
        }
    }
    else {
        for (const channel of channels) {
            subClient.ssubscribe(channel);
        }
        subClient.on("smessageBuffer", (channel, payload) => {
            if (channels.includes(channel.toString())) {
                listener(payload);
            }
        });
    }
}
exports.SSUBSCRIBE = SSUBSCRIBE;
function parseNumSubResponse(res) {
    return parseInt(res[1], 10);
}
function sumValues(values) {
    return values.reduce((acc, val) => acc + val, 0);
}
function PUBSUB(redisClient, arg, channel) {
    if (redisClient.constructor.name === "Cluster" || redisClient.isCluster) {
        // ioredis cluster
        return Promise.all(redisClient.nodes().map((node) => {
            return node
                .send_command("PUBSUB", [arg, channel])
                .then(parseNumSubResponse);
        })).then(sumValues);
    }
    else if (isRedisV4Client(redisClient)) {
        const isCluster = Array.isArray(redisClient.masters);
        if (isCluster) {
            // redis@4 cluster
            const nodes = redisClient.masters;
            return Promise.all(nodes.map((node) => {
                return node.client
                    .sendCommand(["PUBSUB", arg, channel])
                    .then(parseNumSubResponse);
            })).then(sumValues);
        }
        else {
            // redis@4 standalone
            return redisClient
                .sendCommand(["PUBSUB", arg, channel])
                .then(parseNumSubResponse);
        }
    }
    else {
        // ioredis / redis@3 standalone
        return new Promise((resolve, reject) => {
            redisClient.send_command("PUBSUB", [arg, channel], (err, numSub) => {
                if (err)
                    return reject(err);
                resolve(parseNumSubResponse(numSub));
            });
        });
    }
}
exports.PUBSUB = PUBSUB;
